<a href="https://colab.research.google.com/github/HendEmad/Udacity-FWD/blob/main/Explore_old_car_price_data.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

1- Import data from module

In [ ]:
#Importing libraries
import pandas as pd 
import numpy as np

In [ ]:
#Loading data into a data frame
cols = ['symboling', 'normalized-losses', 'make', 'fuel-type', 'aspiration', 'num-of-doors', 'body-style', 'drive-wheels', 'engine-location', 'wheel-base', 'length', 'width', 'height', 'curb-weight', 'engine-type', 'num-of-cylinders', 'engine-size', 'fuel-system', 'bore', 'stroke', 'compression-ratio', 'horsepower', 'peak-rpm', 'city-mpg', 'highway-mpg', 'price']
data = pd.read_csv("/content/imports-85.data.txt", names = cols)
data.shape

(205, 26)

In [ ]:
data.head()


symboling normalized-losses         make  ... city-mpg highway-mpg  price
0          3                 ?  alfa-romero  ...       21          27  13495
1          3                 ?  alfa-romero  ...       21          27  16500
2          1                 ?  alfa-romero  ...       19          26  16500
3          2               164         audi  ...       24          30  13950
4          2               164         audi  ...       18          22  17450

[5 rows x 26 columns]

### ***2- Data Wrangling***
We can see several question marks appeared in the dataframe; those are missing values which may hinder our further analysis.

Steps for working with missing data:
* dentify missing data
* deal with missing data
* correct data format

***2.1 Identify missing values*** 


In [ ]:
# Convert "?" to NaN
data = data.replace("?", np.NaN)
data.head()

symboling normalized-losses         make  ... city-mpg highway-mpg  price
0          3               NaN  alfa-romero  ...       21          27  13495
1          3               NaN  alfa-romero  ...       21          27  16500
2          1               NaN  alfa-romero  ...       19          26  16500
3          2               164         audi  ...       24          30  13950
4          2               164         audi  ...       18          22  17450

[5 rows x 26 columns]

In [ ]:
# Detect missing data
data.isnull().any().any()
# The output is a boolean value indicating whether the value is passed into the argument is in fact missing data.
# "True" stands for missing value, while "False" stands for not missing value.

True

In [ ]:
# Count missing values in each column
data.isnull().sum()

symboling             0
normalized-losses    41
make                  0
fuel-type             0
aspiration            0
num-of-doors          2
body-style            0
drive-wheels          0
engine-location       0
wheel-base            0
length                0
width                 0
height                0
curb-weight           0
engine-type           0
num-of-cylinders      0
engine-size           0
fuel-system           0
bore                  4
stroke                4
compression-ratio     0
horsepower            2
peak-rpm              2
city-mpg              0
highway-mpg           0
price                 4
dtype: int64

Based on the summary above, each column has 205 rows of data, seven columns containing missing data:

* "normalized-losses": 41 missing data
* "num-of-doors": 2 missing data
* "bore": 4 missing data
* "stroke" : 4 missing data
* "horsepower": 2 missing data
* "peak-rpm": 2 missing data
* "price": 4 missing data

***Deal with missing data:***
* drop data
  * drop the whole row
  * drop the whole column
* replace data
  * replace it by mean
  * replace it by frequency
  * replace it based on other functions

***Replace by mean:***
* "normalized-losses": 41 missing data, replace them with mean
* "stroke": 4 missing data, replace them with mean
* "bore": 4 missing data, replace them with mean
* "horsepower": 2 missing data, replace them with mean
* "peak-rpm": 2 missing data, replace them with mean

***Replace by frequency:***
* "num-of-doors": 2 missing data, replace them with "four".
   * Reason: 84% sedans is four doors. Since four doors is most frequent, it is most likely to occur.

***Drop the whole row:***
* "price": 4 missing data, simply delete the whole row
  * Reason: price is what we want to predict. Any data entry without price data cannot be used for prediction; therefore any row now without price data is not useful to us

In [ ]:
# Calculate the average of the column
avg_norm_loss = data['normalized-losses'].astype("float").mean()
avg_norm_loss

122.0

In [ ]:
# Replace "NaN" by mean value
data["normalized-losses"].replace(np.NaN, avg_norm_loss, inplace = True)
#Bore Values
avg_bore = data['bore'].astype('float').mean()
data['bore'].replace(np.NaN, avg_bore, inplace = True)
#Stroke Values
avg_stroke = data['stroke'].astype('float').mean(axis = 0)
data['stroke'].replace(np.nan, avg_stroke, inplace = True)
#horsepower Values 
avg_horsepower = data['horsepower'].astype('float').mean(axis=0)
data['horsepower'].replace(np.nan, avg_horsepower, inplace=True)
#Peak RPM 
avg_peakrpm= data['peak-rpm'].astype('float').mean(axis=0)
data['peak-rpm'].replace(np.nan, avg_peakrpm, inplace=True)

To see which values are present in a particular column, we can use the ".value_counts()" method:

In [ ]:
data['num-of-doors'].value_counts()

four    114
two      89
Name: num-of-doors, dtype: int64

We can see that four doors are the most common type. We can also use the ".idxmax()" method to calculate for us the most common type automatically:

In [ ]:
data['num-of-doors'].value_counts().idxmax()


'four'

In [ ]:
#replace the missing 'num-of-doors' values by the most frequent 
data["num-of-doors"].replace(np.nan, "four", inplace=True)

In [ ]:
# simply drop whole row with NaN in "price" column
data.dropna(subset = ['price'], axis = 0, inplace = True)
data.shape

(201, 26)

In [ ]:
data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 201 entries, 0 to 204
Data columns (total 26 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   symboling          201 non-null    int64  
 1   normalized-losses  201 non-null    object 
 2   make               201 non-null    object 
 3   fuel-type          201 non-null    object 
 4   aspiration         201 non-null    object 
 5   num-of-doors       201 non-null    object 
 6   body-style         201 non-null    object 
 7   drive-wheels       201 non-null    object 
 8   engine-location    201 non-null    object 
 9   wheel-base         201 non-null    float64
 10  length             201 non-null    float64
 11  width              201 non-null    float64
 12  height             201 non-null    float64
 13  curb-weight        201 non-null    int64  
 14  engine-type        201 non-null    object 
 15  num-of-cylinders   201 non-null    object 
 16  engine-size        201 non

In [ ]:
# Convert to proper datatypes
data[["bore", "stroke"]] = data[["bore", "stroke"]].astype("float")
data[["normalized-losses"]] = data[["normalized-losses"]].astype("int")
data[["price"]] = data[["price"]].astype("float")
data[["peak-rpm"]] = data[["peak-rpm"]].astype("float")
data.head()

symboling  normalized-losses         make  ... city-mpg highway-mpg    price
0          3                122  alfa-romero  ...       21          27  13495.0
1          3                122  alfa-romero  ...       21          27  16500.0
2          1                122  alfa-romero  ...       19          26  16500.0
3          2                164         audi  ...       24          30  13950.0
4          2                164         audi  ...       18          22  17450.0

[5 rows x 26 columns]

###***3- Data Standardization***


Data is usually collected from different agencies with different formats. (Data Standardization is also a term for a particular type of data normalization, where we subtract the mean and divide by the standard deviation)

***Standardization :*** Standardization is the process of transforming data into a common format which allows the researcher to make the meaningful comparison.

 ***Example:*** Transform mpg to L/100km.

 In our dataset, the fuel consumption columns "city-mpg" and "highway-mpg" are represented by mpg (miles per gallon) unit. Assume we are developing an application in a country that accept the fuel consumption with L/100km standard

The formula for unit conversion is: 
L/100km = 235 / mpg

In [ ]:
data.head()

symboling  normalized-losses         make  ... city-mpg highway-mpg    price
0          3                122  alfa-romero  ...       21          27  13495.0
1          3                122  alfa-romero  ...       21          27  16500.0
2          1                122  alfa-romero  ...       19          26  16500.0
3          2                164         audi  ...       24          30  13950.0
4          2                164         audi  ...       18          22  17450.0

[5 rows x 26 columns]

In [ ]:
# transform mpg to L/100km by mathematical operation (235 divided by mpg)
data["highway-mpg"] = 235/data["highway-mpg"]
# rename column name from "highway-mpg" to "highway-L/100km"
data.rename(columns = {'highway-mpg':'highway-L/100km'}, inplace=True)

data.head()

symboling  normalized-losses  ... highway-L/100km    price
0          3                122  ...        8.703704  13495.0
1          3                122  ...        8.703704  16500.0
2          1                122  ...        9.038462  16500.0
3          2                164  ...        7.833333  13950.0
4          2                164  ...       10.681818  17450.0

[5 rows x 26 columns]